pkg/hubble/parser/sock/parser.go

```diff
@@ -0,0 +1,217 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Hubble
+
+package sock
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+	"github.com/cilium/cilium/pkg/byteorder"
+	"github.com/cilium/cilium/pkg/hubble/parser/common"
+	"github.com/cilium/cilium/pkg/hubble/parser/errors"
+	"github.com/cilium/cilium/pkg/hubble/parser/getters"
+	"github.com/cilium/cilium/pkg/monitor"
+	monitorAPI "github.com/cilium/cilium/pkg/monitor/api"
+)
+
+// Parser is a parser for SockTraceNotify payloads
+type Parser struct {
+	log            logrus.FieldLogger
+	endpointGetter getters.EndpointGetter
+	identityGetter getters.IdentityGetter
+	dnsGetter      getters.DNSGetter
+	ipGetter       getters.IPGetter
+	serviceGetter  getters.ServiceGetter
+	cgroupGetter   getters.PodMetadataGetter
+	epResolver     *common.EndpointResolver
+}
+
+// New creates a new parser
+func New(log logrus.FieldLogger,
+	endpointGetter getters.EndpointGetter,
+	identityGetter getters.IdentityGetter,
+	dnsGetter getters.DNSGetter,
+	ipGetter getters.IPGetter,
+	serviceGetter getters.ServiceGetter,
+	cgroupGetter getters.PodMetadataGetter,
+) (*Parser, error) {
+	return &Parser{
+		log:            log,
+		endpointGetter: endpointGetter,
+		identityGetter: identityGetter,
+		dnsGetter:      dnsGetter,
+		ipGetter:       ipGetter,
+		serviceGetter:  serviceGetter,
+		cgroupGetter:   cgroupGetter,
+		epResolver:     common.NewEndpointResolver(log, endpointGetter, identityGetter, ipGetter),
+	}, nil
+}
+
+// Decode takes a raw trace sock event payload obtained from the perf event ring
+// buffer and decodes it into a flow
+func (p *Parser) Decode(data []byte, decoded *flowpb.Flow) error {
+	if len(data) == 0 {
+		return errors.ErrEmptyData
+	}
+
+	eventType := data[0]
+	if eventType != monitorAPI.MessageTypeTraceSock {
+		return errors.NewErrInvalidType(eventType)
+	}
+
+	sock := &monitor.TraceSockNotify{}
+	if err := binary.Read(bytes.NewReader(data), byteorder.Native, sock); err != nil {
+		return fmt.Errorf("failed to parse sock trace event: %w", err)
+	}
+
+	isRevNat := decodeRevNat(sock.XlatePoint)
+	ipVersion := decodeIPVersion(sock.Flags)
+
+	dstIP := sock.IP()
+	dstPort := sock.DstPort
+	srcIP := p.decodeSrcIP(sock.CgroupId, ipVersion)
+	srcPort := uint16(0) // source port is not known for TraceSock events
+
+	srcEndpoint := p.epResolver.ResolveEndpoint(srcIP, 0)
+	dstEndpoint := p.epResolver.ResolveEndpoint(dstIP, 0)
+
+	// On the reverse path, source and destination IP of the packet are reversed
+	if isRevNat {
+		srcIP, dstIP = dstIP, srcIP
+		srcPort, dstPort = dstPort, srcPort
+		srcEndpoint, dstEndpoint = dstEndpoint, srcEndpoint
+	}
+
+	decoded.Verdict = flowpb.Verdict_TRANSLATED
```

Comment:
We should set this only on the post_* notifications. This isn't applicable for pre_* events.

Reply:
@aditighag PTAL. I've changed the code to use TRANSLATED for post events and the newly introduced TRACED verdict for pre events (as discussed with the Hubble team a week ago, we want a verdict here. Verdict UNKNOWN will be confusing to users, since it indicates missing data):

https://github.com/cilium/cilium/blob/c5372514225a0335457d65fb66fdec8b72cf03b3/pkg/hubble/parser/sock/parser.go#L215-L226

I've also added a short comment to each verdict value to document what it does.